pkg/cloudprovider/providers/qingcloud/qingcloud_instances.go

```diff
@@ -0,0 +1,183 @@
+
+	ins, err := qc.getInstanceById(nodeNameToInstanceId(nodeName))
+	if err != nil {
+		return "", nil
+	}
+
+	return ins.InstanceID, nil
+}
+
+// InstanceType returns the type of the specified instance.
+func (qc *Qingcloud) InstanceType(name types.NodeName) (string, error) {
+	glog.V(3).Infof("InstanceType(%v) called", name)
+
+	ins, err := qc.getInstanceById(nodeNameToInstanceId(name))
+	if err != nil {
+		return "", err
+	}
+
+	return ins.InstanceType, nil
+}
+
+// List lists instances that match 'filter' which is a regular expression which must match the entire instance name (fqdn)
+func (qc *Qingcloud) List(filter string) ([]types.NodeName, error) {
+	glog.V(3).Infof("List(%v) called", filter)
+
+	instances, err := qc.getInstancesByFilter(filter)
+	if err != nil {
+		glog.Errorf("error getting instances by filter '%s': %v", filter, err)
+		return nil, err
+	}
+	result := []types.NodeName{}
+	for _, ins := range instances {
+		result = append(result, types.NodeName(ins.InstanceID))
+	}
+
+	glog.V(3).Infof("List instances: %v", result)
+
+	return result, nil
+}
+
+// AddSSHKeyToAllInstances adds an SSH public key as a legal identity for all instances.
+// The method is currently only used in gce.
+func (qc *Qingcloud) AddSSHKeyToAllInstances(user string, keyData []byte) error {
+	return errors.New("Unimplemented")
+}
+
+// CurrentNodeName returns the name of the node we are currently running on
+// On most clouds (e.g. GCE) this is the hostname, so we provide the hostname
+func (qc *Qingcloud) CurrentNodeName(hostname string) (types.NodeName, error) {
+	return types.NodeName(hostname), nil
+}
+
+func (qc *Qingcloud) getInstanceById(instanceId string) (*instance.Instance, error) {
+	instanceN := qingcloud.NumberedString{}
+	instanceN.Add(instanceId)
+	statusN := qingcloud.NumberedString{}
+	statusN.Add("pending", "running", "stopped")
+	verbose := qingcloud.Integer{}
+	verbose.Set(1)
+	resp, err := qc.instanceClient.DescribeInstances(instance.DescribeInstanceRequest{
+		InstancesN: instanceN,
+		StatusN:    statusN,
+		Verbose:    verbose,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.InstanceSet) == 0 {
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	return &resp.InstanceSet[0], nil
+}
+
+// List instances that match the filter
+func (qc *Qingcloud) getInstancesByFilter(filter string) ([]instance.Instance, error) {
+	searchWord := qingcloud.String{}
+	searchWord.Set(filter)
+	statusN := qingcloud.NumberedString{}
+	statusN.Add("running", "stopped")
+	verbose := qingcloud.Integer{}
+	verbose.Set(1)
+	limit := qingcloud.Integer{}
+	limit.Set(100)
+
+	instances := []instance.Instance{}
+
+	for i := 0; ; i += 100 {
+		offset := qingcloud.Integer{}
+		offset.Set(i)
+		resp, err := qc.instanceClient.DescribeInstances(instance.DescribeInstanceRequest{
+			SearchWord: searchWord,
+			StatusN:    statusN,
+			Verbose:    verbose,
+			Offset:     offset,
+			Limit:      limit,
+		})
+		if err != nil {
+			return nil, err
```

Comment:
up to here, there are a partial list, would it be usable? I am not familiar with qingcloud (and its api) error handling, on this one and other comments on retries and garbage collection, I would like to see api able to handle exceptions.